apps/src/dance/Dance.js

```diff
@@ -118,6 +119,28 @@ Dance.prototype.init = function (config) {
     config.valueTypeTabShapeMap = {[Blockly.BlockValueType.SPRITE]: 'angle'};
 
     this.studioApp_.init(config);
+    this.currentCode = this.studioApp_.getCode();
+    if (this.usesPreview) {
+      this.studioApp_.addChangeHandler(e => {
+        // We want to check if the workspace code changed only when a block has been moved or
+        // if a block has changed.
+        // A move event is fired when a block is dragged and then dropped.
+        if (
+          e.type !== Blockly.Events.BLOCK_MOVE &&
+          e.type !== Blockly.Events.BLOCK_CHANGE
+        ) {
+          return;
```

Comment:
Is this still necessary given the below check for code changing?  Or is it an optimization to avoid having to do the below check?

Reply:
Yes - this is still necessary. In the screencast video below, a user drags out a 'Make new dance' block which triggers a change in the workspace and the new code is not the same as the current code. The command `makeAnonymousDanceSprite` is generated and error results (seen in dev console).

https://github.com/code-dot-org/code-dot-org/assets/107423305/b65ec3e4-9cfa-4435-a2af-5ad1e08e6aef

This error could be fixed by making a small update to the `dance-party` repo, but then a gray sprite would appear until the block was moved to under the `setup` block. See first screencast video in [this comment](https://github.com/code-dot-org/code-dot-org/pull/53817#discussion_r1330854259).

Here is a [list](https://github.com/google/blockly/blob/9089661f9d2209fe783a54c058e1ecaa7b8bcf7f/core/events/events.ts#L99-L129) of Blockly events. I opted to update the preview if only the `MOVE` or `CHANGE` event was triggered. More details in the PR description. Here's a screencast with the dev console open.

https://github.com/code-dot-org/code-dot-org/assets/107423305/4e5cea4a-2191-43e7-bc53-b30d50b25c93

Reply:
@breville noted that is was interesting that in preview mode, before I added the events check for 'move' and 'change', that the 'make new dance' code was executed as it was being dragged onto the workspace but not actually part of the  'setup' block if I made a tweak in the `dance-repo` code to avoid the error.  In other words, that the call to `preview` is compiling the JS and then executing the code for that block, even though it's not yet attached to setup.

https://github.com/code-dot-org/code-dot-org/assets/107423305/9ba3f3dd-81ad-4087-99cd-6c8356a96b99

Just wanted to note my learnings as I looked into his question. 
Here's a screencast of the block being dragged out (when `usesPreview` is set to `false`:

https://github.com/code-dot-org/code-dot-org/assets/107423305/45315aa7-61e2-47ba-a027-111f455c317d

In `Dance.js`, we see:
```
this.hooks.find(v => v.name === 'runUserSetup').func();
```
In `p5.dance.interpreted.js` in `dance-party` repo, the function `runUserSetup` is defined [here](https://github.com/code-dot-org/dance-party/blob/main/src/p5.dance.interpreted.js).
`setupCallbacks` is an array defined in same file. Callbacks are added with `registerSetup`, `whenSetup`, and `whenSetupSong` (block only used in /allthethings dance levels).
`registerSetup` is what I’m curious about. It’s a constant defined as part of `DancelabReservedWords` in `dance/constants.js`.
The other  thing to note is that when the callback func is executed, this `func` is [defined](https://github.com/code-dot-org/dance-party/blob/33415baafae75b12fd712dcf0cfc7013751d48b0/src/p5.dance.js#L14) in a `Behavior`. In this case, the `func` was [assigned ](https://github.com/code-dot-org/dance-party/blob/33415baafae75b12fd712dcf0cfc7013751d48b0/src/p5.dance.js#L438)`updateSpriteFrame`.
So still an open question about how code not in a `setup` block is pushed to the `setupCallbacks` array…maybe via the mysterious `registerSetup`?

Reply:
The PR that first added `p5.dance.interpreted` https://github.com/code-dot-org/dance-party/pull/49